src/EditorFeatures/Core/Implementation/CodeFixes/CodeFixService.cs

```diff
@@ -83,32 +83,101 @@ internal partial class CodeFixService : ForegroundThreadAffinitizedObject, ICode
-                    // at the same time, let the task to run to finish. if the lightbulb is expanded, we just simply use the task to get all fixes.
-                    //
-                    // first approach is simpler, so I will implement that first. if the first approach turns out to be not good enough, then
-                    // I will try the second approach which will be more complex but quicker
-                    var hasFix = await ContainsAnyFix(document, diagnostic, cancellationToken).ConfigureAwait(false);
-                    if (hasFix)
+                var results = new List<DiagnosticData>();
+                try
+                {
+                    while (taskSet.Count > 0)
                     {
-                        return new FirstDiagnosticResult(!fullResult, hasFix, diagnostic);
+                        var completedTask = await Task.WhenAny(taskSet).ConfigureAwait(false);
+                        if (completedTask == errorDiagnosticTask && completedTask.Result != null)
+                        {
+                            // Error task returned a result; just return it.
+                            return CreateResult(fullResult, completedTask.Result);
+                        }
+                        else if (errorDiagnosticTask == null ||
+                            errorDiagnosticTask.IsCompleted && errorDiagnosticTask.Result == null)
+                        {
+                            // Some other task completed and we know there's no fixable error diagnostic.
+                            return CreateResult(fullResult, completedTask.Result);
+                        }
+                        else
+                        {
+                            // We don't know if we're going to show errors yet. Keep this result.
+                            results.Add(completedTask.Result);
+                        }
+
+                        taskSet.Remove(completedTask);
                     }
                 }
+                finally
+                {
+                    linkedTokenSource.Cancel();
+                }
+
+                // Since we treat produce the same light bulb category for all
+                // non-error diagnostics, just use the first one.
+                return CreateResult(fullResult, results.FirstOrDefault());
+            }
+        }
+
+        private Task<DiagnosticData> AddErrorDiagnosticTask(
+            Document document,
+            TextSpan range,
+            IEnumerable<IGrouping<DiagnosticSeverity, DiagnosticData>> diagnosticGroups, 
+            ISet<Task<DiagnosticData>> taskSet,
+            CancellationToken cancellationToken)
+        {
+            var errorDiagnosticGroup = diagnosticGroups.FirstOrDefault(d => d.Key == DiagnosticSeverity.Error);
+            if (errorDiagnosticGroup != null)
+            {
+                var task = Task.Factory.SafeStartNewFromAsync(
```

Comment:
1. just use Task.Run.  
2. just return this task and add it to the taskset from the call if not null (i think we have an AddIfNotNull extension for just that).